File: backend/app/catalog_view/magic.py
"""Magic and resonance: powers, spells, spirits, foci, and their echoes."""

from __future__ import annotations

from ..data_loader import CatalogDict


def section(raw: CatalogDict) -> dict:
    return {
        "powers": [
            {
                "id": p["id"],
                "name": p["name"],
                "points": p["points"],
                "levels": p["levels"],
                "maxlevels": p["maxlevels"],
                "extrapointcost": p["extrapointcost"],
                "source": p["source"],
                "page": p["page"],
                "select": p.get("select"),
                "required": list(p.get("required") or []),
                "adeptway": p.get("adeptway") or 0,
                "adeptwayrequires": list(p.get("adeptwayrequires") or []),
            }
            for p in raw.get("powers") or []
            if not p.get("hidden")
        ],
        "enhancements": [
            {
                "id": e["id"],
                "name": e["name"],
                "power": e.get("power"),
                "source": e.get("source"),
                "page": e.get("page"),
                "required": e.get("required") or {},
            }
            for e in raw.get("enhancements") or []
        ],
        "mentors": [
            {
                "id": m["id"],
                "name": m["name"],
                "source": m.get("source"),
                "page": m.get("page"),
                "advantage": m.get("advantage") or "",
            }
            for m in raw.get("mentors") or []
        ],
        "spells": [
            {
                "id": s["id"],
                "name": s["name"],
                "category": s.get("category"),
                "dv": s.get("dv"),
                "type": s.get("type"),
                "range": s.get("range"),
                "duration": s.get("duration"),
                "descriptor": s.get("descriptor"),
                "kind": s.get("kind") or "spell",
                "useskill": s.get("useskill") or "Spellcasting",
                "learnable": bool(s.get("learnable")),
                "required": [name for names in (s.get("required") or {}).values() for name in names],
                "source": s.get("source"),
                "page": s.get("page"),
            }
            for s in raw.get("spells") or []
        ],
        "traditions": [
            {
                "id": t["id"],
                "name": t["name"],
                "drain": t.get("drain") or "",
                "drain_attrs": list(t.get("drain_attrs") or []),
                "spirits": dict(t.get("spirits") or {}),
                "source": t.get("source"),
                "page": t.get("page"),
            }
            for t in raw.get("traditions") or []
        ],
        "spirits": [
            {
                "id": s["id"],
                "name": s["name"],
                "attributes": dict(s.get("attributes") or {}),
                "powers": list(s.get("powers") or []),
                "optionalpowers": list(s.get("optionalpowers") or []),
                "skills": list(s.get("skills") or []),
                "weaknesses": list(s.get("weaknesses") or []),
                "source": s.get("source"),
                "page": s.get("page"),
            }
            for s in raw.get("spirits") or []
        ],
        "complex_forms": [
            {
                "id": f["id"],
                "name": f["name"],
                "target": f.get("target") or "",
                "duration": f.get("duration") or "",
                "fv": f.get("fv") or "",
                "needs_extra": bool(f.get("needs_extra")),
                "required": [name for names in (f.get("required") or {}).values() for name in names],
                "source": f.get("source"),
                "page": f.get("page"),
            }
            for f in raw.get("complex_forms") or []
        ],
        "streams": [
            {
                "id": s["id"],
                "name": s["name"],
                "drain": s.get("drain") or "",
                "drain_attrs": list(s.get("drain_attrs") or []),
                "sprites": list(s.get("sprites") or []),
                "source": s.get("source"),
                "page": s.get("page"),
            }
            for s in raw.get("streams") or []
        ],
        "sprites": [
            {
                "id": s["id"],
                "name": s["name"],
                "attributes": dict(s.get("attributes") or {}),
                "powers": list(s.get("powers") or []),
                "skills": list(s.get("skills") or []),
                "source": s.get("source"),
                "page": s.get("page"),
            }
            for s in raw.get("sprites") or []
        ],
        "foci": [
            {
                "id": f["id"],
                "name": f["name"],
                "maxrating": f.get("maxrating") or 6,
                "cost": f.get("cost") or "",
                "effect": f.get("effect") or "",
                "formula": (
                    {
                        "id": (f.get("formula") or {}).get("id"),
                        "name": (f.get("formula") or {}).get("name"),
                        "cost": (f.get("formula") or {}).get("cost") or "",
                    }
                    if f.get("formula")
                    else None
                ),
                "source": f.get("source"),
                "page": f.get("page"),
            }
            for f in raw.get("foci") or []
        ],
        "qi_focus": raw.get("qi_focus"),
        "metamagics": [
            {
                "id": item["id"],
                "name": item["name"],
                "adept": bool(item.get("adept")),
                "magician": bool(item.get("magician")),
                "repeatable": bool(item.get("repeatable")),
                "required": [name for names in (item.get("required") or {}).values() for name in names],
                "source": item.get("source") or "",
                "page": item.get("page") or "",
            }
            for item in raw.get("metamagics") or []
        ],
        "magic_arts": [
            {
                "id": item["id"],
                "name": item["name"],
                "source": item.get("source") or "",
                "page": item.get("page") or "",
            }
            for item in raw.get("magic_arts") or []
        ],
        "echoes": [
            {
                "id": item["id"],
                "name": item["name"],
                "max_takes": item.get("max_takes"),
                "needs_extra": bool(item.get("needs_extra")),
                "source": item.get("source") or "",
                "page": item.get("page") or "",
            }
            for item in raw.get("echoes") or []
        ],
    }

File: frontend/lib/types/rows/magic.ts
// Magic and resonance: row shapes listed in `Derived` (../derived.ts).

import type { MentorChoice } from "../installs";
import type { Notice } from "@/lib/engine-notices";

export interface InstalledSpell {
  id: string;
  spell_id: string;
  name: string;
  category?: string;
  kind?: "spell" | "ritual" | "enchantment";
  useskill?: string;
  has_force?: boolean;
  type?: string;
  range?: string;
  duration?: string;
  descriptor?: string;
  dv: string;
  damage?: string;
  damage_mod?: number;
  required?: string[];
  source?: string;
  page?: string;
  free: boolean;
  karma: number;
  barehanded_adept?: boolean;
  alchemical?: boolean;
  granted?: boolean;
  focus_bonus?: number;
  spell?: SpellCastInfo | null;
}

export interface InstalledComplexForm {
  id: string;
  form_id: string;
  name: string;
  label?: string;
  target: string;
  duration: string;
  fv: string;
  extra?: string;
  needs_extra?: boolean;
  options?: string[];
  level: number;
  level_min: number;
  level_max: number;
  fade: number | null;
  fade_code?: "S" | "P" | null;
  physical?: boolean;
  resist: number;
  resist_attrs: string;
  free: boolean;
  karma: number;
  test?: MagicTestInfo;
  source?: string;
  page?: string;
}

/** One power a spirit or sprite comes with, and what SR5 p.394 gives it.
 *  `type` is `M` / `P`; the rest are the words the data file uses (`Complex`,
 *  `LOS`, `Sustained`, …), or a formula the table spells out. */
export interface CritterPower {
  name: string;
  type?: string;
  action?: string;
  range?: string;
  duration?: string;
  source?: string;
  page?: string;
}

export interface InstalledSprite {
  id: string;
  sprite_id: string;
  name: string;
  level: number;
  level_max: number;
  services: number;
  registered?: boolean;
  hits?: number | null;
  opposed_hits?: number | null;
  test?: MagicTestInfo;
  attributes?: Record<string, number>;
  matrix?: {
    attack: number;
    sleaze: number;
    dataprocessing: number;
    firewall: number;
    initiative: number;
  };
  powers?: CritterPower[];
  skills?: { name: string; attribute?: string; rating: number }[];
  source?: string;
  page?: string;
}

export interface TraditionInfo {
  id: string;
  name: string;
  drain: string;
  drain_attrs: string[];
  spirits?: Record<string, string>;
  source?: string;
  page?: string;
}

export interface MagicTestInfo {
  skill: string;
  rating: number;
  attr: string;
  attr_value: number;
  bonus: number;
  pool: number;
  defaulted?: boolean;
  missing?: boolean;
  force: number;
  limit: number;
  limit_name?: string;
  vs: number;
  hits?: number | null;
  opposed_hits?: number | null;
  net?: number | null;
  drain?: number | null;
  drain_code?: "S" | "P" | null;
  physical?: boolean;
  days?: number | null;
}

export interface InstalledSpirit {
  id: string;
  spirit_id: string;
  name: string;
  role?: string;
  role_label?: Notice | null;
  force: number;
  force_max: number;
  services: number;
  nuyen: number;
  bound?: boolean;
  hits?: number | null;
  opposed_hits?: number | null;
  test?: MagicTestInfo;
  attributes?: Record<string, number>;
  powers?: CritterPower[];
  optionalpowers?: CritterPower[];
  skills?: { name: string; attribute?: string; rating: number }[];
  weaknesses?: string[];
  source?: string;
  page?: string;
}

export interface InstalledFocus {
  id: string;
  gear_id: string;
  name: string;
  force: number;
  force_max: number;
  nuyen: number;
  karma: number;
  crafted?: boolean;
  formula_bought?: boolean;
  formula_nuyen?: number;
  reagent_nuyen?: number;
  retail_nuyen?: number;
  hits?: number | null;
  opposed_hits?: number | null;
  test?: MagicTestInfo;
  formula_test?: MagicTestInfo;
  effect?: string;
  formula?: { id?: string; name?: string; cost?: string } | null;
  needs_weapon?: boolean;
  weapon_type?: string;
  weapon_id?: string;
  weapon_name?: string;
  weapon_dice?: number;
  weapon_options?: { id: string; name: string }[];
  source?: string;
  page?: string;
}

export interface InstalledAdeptPower {
  id: string;
  power_id: string;
  name: string;
  rating: number;
  total_rating?: number;
  free_levels?: number;
  rating_min: number;
  rating_max: number;
  extra: string;
  cost: number;
  full_cost?: number;
  discounted?: boolean;
  can_discount?: boolean;
  select?: "skill" | "attribute" | "spell" | null;
  options: string[];
  source?: string;
  page?: string;
  notes?: Notice[];
  free_only?: boolean;
  spell?: SpellCastInfo | null;
}

export interface SpellCastInfo {
  spell_id: string;
  name: string;
  category?: string;
  type?: string;
  range?: string;
  duration?: string;
  descriptor?: string;
  dv: string;
  damage?: string;
  damage_mod?: number;
  drain_mod?: number;
  force: number;
  force_min: number;
  force_max: number;
  drain: number | null;
  drain_code: "S" | "P" | null;
  physical: boolean;
  resist: number;
  resist_attrs: string;
  barehanded_adept?: boolean;
}

export interface InstalledQiFocus {
  id: string;
  rating: number;
  rating_min: number;
  rating_max: number;
  power_id: string;
  name: string;
  power_rating: number;
  power_rating_max: number;
  extra: string;
  select?: "skill" | "attribute" | "spell" | null;
  options: string[];
  nuyen: number;
  karma: number;
  source?: string;
}

export interface MentorInfo {
  id: string;
  name: string;
  advantage: string;
  disadvantage: string;
  source?: string;
  choices: MentorChoice[];
}

export interface EnhancementInfo {
  id: string;
  name: string;
  power?: string | null;
  karma: number;
  source?: string;
  page?: string;
  ok?: boolean;
}
